cpp/src/io/utilities/wrapper_utils.hpp

```diff
@@ -146,3 +146,53 @@ struct rmm_deleter {
 };
 template <typename T>
 using device_ptr = std::unique_ptr<T, rmm_deleter<T>>;
+
+/**
+ * @brief A helper class that owns a resizable device memory buffer.
+ *
+ * Copy construction and copy assignment are disabled to prevent
+ * accidental copies.
+ **/
+template <typename T>
+class device_buffer {
+  T* d_data_ = nullptr;
+  size_t count_ = 0;
+  const cudaStream_t stream_ = 0;
+
+public:
+  device_buffer() noexcept = default;
+  device_buffer(size_t cnt, cudaStream_t stream = 0):
+    stream_(stream), count_(cnt){
+    resize(count_);
+  }
+
+  T* data() const noexcept {return d_data_;}
+  size_t size() const noexcept {return count_;}
+
+  void resize(size_t cnt) {
+    count_ = cnt;
+    const auto error = RMM_REALLOC(&d_data_, count_*sizeof(T), stream_);
```

Comment:
I don't think you should use RMM_REALLOC currently, it's broken (doesn't copy the data). (since the underlying allocator doesn't support it, it's hard to implement.)

Reply:
I also don't think it works if you call it before calling RMM_ALLOC. Is that intended to work with other allocators?

Reply:
@harrism 
In the current use cases, we don't need the data to be copied. I'll modify to include the copy, to make the class more generally applicable.

realloc is defined to behave the same as (m)alloc when the input pointer points to null. I check both cudaRealloc() specs and the rmm::realloc code.
Another edge case I checked for is passing zero size to realloc. rmm::realloc sets the pointer to null, which is exactly what I need here.

Reply:
@harrism replace the realloc with malloc+copy, please review.

Reply:
That's better. How did you check `cudaRealloc()` specs? There's no such thing as `cudaRealloc`.

Reply:
Turns out I checked the C++ realloc for reference on what to expect from realloc behavior in general, before digging into the rmm code. Apologies for the confusion (I'll blame it on the late hours).